Check for memory overlap in more ATen ops

peterbell10 opened the pull request:
Fixes #39639 (and a lot more)

I went through every use of `TensorIterator` and added the `check_mem_overlap` flag to all the inplace ops, or ops using `out` tensors. I only touched `TH`/`THC` where the operation was implemented by `TensorIterator` on the other device for consistency. Hopefully the remaining ops can be added as they are ported to `ATen`.

peterbell10 committed:
TensorIterator: Check for memory overlap by default

peterbell10 commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -298,60 +298,63 @@ void bernoulli_scalar_kernel(Tensor& self, double p, c10::optional<Generator> ge
-      // See Note [Acquire lock when using random generators]
-      std::lock_guard<std::mutex> lock(generator->mutex_);
-      seed = generator->random();
+  // FIXME: Is a check really needed here? Should fail to compile if no MKL
+  if (!cpuinfo_initialize()) { AT_ASSERTM(false, "cpuinfo failed to initialize"); }
+  const auto vendor = cpuinfo_get_processor(0)->core->vendor;
+  if (vendor != cpuinfo_vendor_intel && vendor != cpuinfo_vendor_amd) {
```
This doesn't really belong  in this PR but I can't test bernoulli without it. I can't see why we should disallow compiling with MKL on AMD CPUs.

ezyang commented:
One thing I'm not sure about is (1) is it true that we should be universally checking for memory overlap in all inplace/out TensorIterator invocations, and (2) if so, why isn't this built into the API in a more structured way, rather than as an argument to each call site that you could forget.

ezyang commented:
@peterbell10 please add the original author of memory overlap and the reviewers of that PR as reviewers to this PR

peterbell10 commented:
cc @pbelevich who seems to have done most of the TensorIterator work in #22917 and #24058

>is it true that we should be universally checking for memory overlap in all inplace/out TensorIterator invocations

Most of the time, I think so. There are some cases like reductions where the tensors are intentionally given 0-strides and so these checks would give false-positives.

>why isn't this built into the API in a more structured way, rather than as an argument to each call site that you could forget.

Perhaps we should remove the defaults from the `n-ary_op` functions. People creating custom tensor iterators could still forget to call `set_check_mem_overlap` though.

ezyang commented:
Removing myself as reviewer.

peterbell10 committed:
Fix distribution tests

Actual bug in distributions being caught by the new checks here.

peterbell10 commented:
As a side note, in `numpy` broadcasted arrays are marked as read-only to avoid these issues:

```python
>>> x = np.random.rand(100)
>>> y = np.broadcast_to(x, (10, 100))
>>> y[0, 0] = 1
ValueError: assignment destination is read-only
>>> y.flags['WRITEABLE']
False
```

zou3519 commented:
> Perhaps we should remove the defaults from the n-ary_op functions. 

I'm in favor of removing the `check_mem_overlap` argument from `TensorIterator::nullary_op` / `TensorIterator::binary_op` / `TensorIterator::unary_op` and just having those always check memory overlap. I don't see any cases when it comes to those where we wouldn't want to check memory overlap but it's also been a while since I've looked at the code.

> People creating custom tensor iterators could still forget to call set_check_mem_overlap though.

Something to consider is that we could just default to checking memory overlap in custom tensor iterators but selectively disable it for the cases we know won't work. It sounds like reductions are one such case?

peterbell10 commented:
I've rebased on the latest master, changed the default in `TensorIteratorConfig` to always check for memory overlap and removed the `check_mem_overlap` parameter from the `n-ary_op` constructors.

ezyang commented:
Thanks!

zou3519 commented on torch/distributions/multinomial.py:
```diff
@@ -101,7 +101,8 @@ def sample(self, sample_shape=torch.Size()):
         if self._validate_args:
             self._validate_sample(value)
-        logits, value = broadcast_all(self.logits.clone(memory_format=torch.contiguous_format), value)
+        logits, value = broadcast_all(self.logits, value)
```
Why did this have to change?

peterbell10 replied:
This was a bug exposed by the new checks in `masked_fill`. It first broadcasts `self.logits` with `value`, then writes to the broadcasted (i.e. zero-strided) array in a way that overwrites entire rows instead of only where `value == 0`.

zou3519 commented on test/test_distributions.py:
```diff
@@ -4905,8 +4905,8 @@ def f(*values):
-            expected = f(*values)
-            actual = traced_f(*values)
+            expected = f(*values).clone()
+            actual = traced_f(*values).clone()
```
Why did this have to change?

peterbell10 replied:
The next line writes into a broadcasted array:
`expected[expected == float('inf')] = 0.`

zou3519 replied:
This seems BC-breaking. Even if `expected` is a broadcasted array, `expected[expected == float('inf')] = 0.` produces a correct result because it is setting the right elements to 0

peterbell10 replied:
Note that this is a `masked_fill_` which is exactly the operation in the original issue. To preserve the correct uses, like
```python
expected[expected == float('inf')] = 0.
```
then we could inspect the mask along the broadcasted dimensions and only allow the fill if the mask is constant along broadcasted dimensions. Maybe deprecating that behavior as well.

zou3519 reviewed (commented):
(not a full review)

zou3519 commented:
Hmm, there are some operations where checking for memory overlap isn't necessary and would cause backwards incompatibility. For example, checking memory overlap for something like `torch.tensor(0.4).expand(3).cos_()` is good, because there's a chance that the result is `cos(cos(cos(0.4)))`.

However, we have cases like `torch.tensor(0.4).expand(3).zero_()` where the result of the operation is unambiguous and this would lead to backwards incompatibility. Maybe that's OK. I am trying to brainstorm if there are other cases like this (the `expected[expected == float('inf')] = 0`) example seems like one of them)

peterbell10 commented:
For `zero_` specifically, perhaps fill with scalar could be made an exception. However, I'm not sure what you could be planning to do with it afterwards. Presumably you'd then start writing to indexed locations and then it becomes buggy again. So, I'm unsure that a exception would be helpful.

Perhaps fill should just raise a deprecation warning instead of an error?

peterbell10 committed:
Fix unfold_backward

zou3519 commented:
@peterbell10 sorry, this fell off my radar -- I'll have more time to review this starting next week and will give this a read through then

zou3519 commented:
cc @peterbell10 I thought about the problem we’re addressing here and wrote up some comments. Please let me know what your thoughts are.

Also cc @mruberry (for numpy compatibility) and @VitalyFedyunin @ngimel @ezyang (for TensorIterator). I'm curious if you folks have thoughts on the internal memory overlap problem and potential solutions.

## What is the issue we’re trying to solve?

As a common pitfall, users will expand tensors and then perform in-place operations on them. We’ve mitigated this problem in the past by adding internal memory overlap checks to some of the more common operations (e.g., we check if a Tensor has stride 0 and size > 1). However, this problem still exists for a lot of other operators.

## My concerns on this Pull Request

**Potential for BC-breaking behavior**: There are some operators where if an output has internal memory overlap, or if an output and input have partial memory overlap, the result is always correct (e.g., `Tensor.zero_`). There are some other more interesting cases, like
```
expected = torch.randn(3).expand(4, 5, 3)
expected[expected == float('inf')] = 0
```
where the output tensor has internal overlap, but it is OK that it does and making a change like this would be BC-breaking (in fact, there is a test in this PR that we had to change because of this). Maybe these cases don’t actually matter, but we I think if we go down the route of this PR, we should take a detailed survey of the operations involved that we would need to deprecate and have a deprecation plan.

**Decision to make TensorIterator check for memory overlap by default.** 

This isn't a complete solution to our problem because not all PyTorch operations use TensorIterator or will use TensorIterator in the future. Secondly, we’ve been doing a lot of TH->ATen migrations. If we change TensorIterator to check for memory overlap by default, then TH->ATen migrations may unwillingly introduce BC-breaking changes.

## Alternatives

I think if we have to go through the trouble of identifying all the BC-breaking operators and deprecating the "expand + modify-in-place" behavior for those operators, we should take some time to explore the problem space. Assuming we have to break BC, it'll be 6-9 months (2-3 releases) to finish deprecating and updating everything.

One alternative that seems nice to me is to introduce a WRITEABLE flag to tensors, a la numpy. Broadcasted tensors (obtained from e.g. `torch.expand`, `torch.broadcast_tensors`) would have WRITEABLE=False and raise an error message on in-place operations. If it is possible to bake this change into the codebase structurally, then it sounds like a better solution: instead of having to check internal memory-overlap over the codebase, maybe we can just error out when an operator tries to ask for the tensor's `data_ptr` if WRITEABLE=False.

peterbell10 commented:
>One alternative that seems nice to me is to introduce a WRITEABLE flag to tensors, a la numpy. [...] instead of having to check internal memory-overlap over the codebase, maybe we can just error out when an operator tries to ask for the tensor's data_ptr if WRITEABLE=False.

To go down this route I think there would need to be `const` and `mutable` variants of all accessors like `data_ptr` (either by different names, or say a `ConstTensor` class). Then each operator would need to be updated for const-correctness. Otherwise, I think you'd end up with functions that only read from a tensor but would none-the-less require the inputs be writeable.

A huge benefit of having `TensorIterator` take care of it is that it already knows which tensors are inputs and which are outputs. I suppose something similar could be done with the code generation since modified tensors are marked with `!`. These checks could be enabled everywhere, but at first only emit only a deprecation warning and then down the line make it a full error. Notably, this solution would cover all native functions and not just those using `TensorIterator`.

ezyang commented:
I'm not exactly sure what to do here. Let me describe some high level things to think about.

1. If we break BC, things are less bad if we have reasonable ways for people to rewrite their code in the new way. `expected[expected == float('inf')] = 0` is a good case to think about: if we make this stop working, is there a way for someone to recover the old behavior? (It doesn't sound like you can conveniently do so in this example; it seems like we need some sort of de-expand operation. But it's worth pointing out that this code is doing more work than it needs to, since `expected == float('inf')` is going to uselessly redo the inf test for the broadcasted elements.)
2. I don't think the test suite is necessarily a good test for what people are likely to do in the wild. We have a lot of synthetic examples to exercise particular cases. So be careful about overoptimizing for the test case.
3. Regarding Peter Bell's suggestion, @smessmer has advocated for `const` and `mutable` variants of `data_ptr` in the past. This was in a different context (cc @ljk53) where we need to update version counters on storages when we do mutable operations on them. It's worth noting that we actually do have the ability to identify what operators do mutations at a higher level (e.g., by just looking at schema), as mutability is annotated in schema. So you don't have to do it `data_ptr` style; and it might be feasible to include test for mutating broadcasting at the same time you check if you need to update a version counter. However, one complicating factor is that you want to error BEFORE you actually mutate anything (whereas version counters should be updated AFTER you mutate something.)
4. When possible, opt for something simpler and requires less work when things are confusing.
5. I would guess there aren't too many operations which work "correctly" when you pass them in broadcasted outputs; few enough that it would be fine to just special case them specifically.

Hope that helps.

peterbell10 committed:
Allow scalar filling of broadcasted tensors

peterbell10 commented:
I've rebased and added an exception for `.fill_`.

>`expected[expected == float('inf')] = 0` is a good case to think about: if we make this stop working, is there a way for someone to recover the old behavior?

With numpy's flags system the user can set the writable flag manually, then write to the array as normal. Though, a more idiomatic way would be to use `where(expected == float('inf'), 0, expected)` instead of writing to the array.

zou3519 commented:
Now that https://github.com/pytorch/pytorch/pull/41923 is closed I am returning to review this.

So from the tests it looks like we've added memory overlap support for the following:
```
random_
uniform_
cauchy_
log_normal_
exponential_
geometric_
normal_
fmod
lerp
bernoulli_/bernoulli
index_put_
masked_fill_
masked_select_
masked_scatter_
index_select_
cat
scatter_
gather_
linspace
logspace
```
Did we enable memory overlap for any other operators? I feel like switching the default on TensorIterator to check memory overlap should have affected more operators

peterbell10 commented:
Most `TensorIterator` operators were already using the `check_mem_overlap=True` option. In an earlier version of this PR I had gone through all `TensorIterator` users and added `check_mem_overlap=True` wherever it was missing and those test cases matched that version.

zou3519 commented:
@peterbell10, sorry for the delay in review. I had a conversation with @mruberry offline about this PR. We came to the following ideas:
- Let's not pursue alternative solutions (like adopting a WRITABLE flag) for now. An alternative solution would require much larger changes that could happen in the future, but we would still like this PR to go in because it bring value and is not in conflict with an alternative solution.
- We should try not to make any previously correct behavior BC-breaking. So we'll want exceptions for `fill_` (already included in this PR)` and other functions like `fill_` (I'll check if there are any)
- Having the default for TensorIterator check memory overlap (what is currently written in this PR) is good because there are more operations that will want to check memory overlap than not.

From those points, I'll review the pull request from the following perspective:
- I'll take an audit of all of the operators touched in the PR, and check for BC-breaking behavior.

mruberry commented:
> * We should try not to make any previously correct behavior BC-breaking. So we'll want exceptions for `fill_` (already included in this PR)`and other functions like`fill_` (I'll check if there are any)

It'd also be nice to document when functions should and shouldn't be setting the check flag.

zou3519 commented on test/test_torch.py:
```diff
@@ -13817,7 +13831,7 @@ def test_ternary_op_mem_overlap(self, device, dtype):
             ("addcdiv", True, True, 'cuda'),
             ("lerp", True, True, 'cpu'),
-            ("lerp", False, False, 'cuda')
+            ("lerp", True, True, 'cuda')
```
I did an audit of `TensorIterator::unary_op`, `TensorIterator::binary_op`, and `TensorIterator::nullary_op` cases in the code that use the default. The default before this PR was check_mem_overlap=False, but after this PR, those are True. I found a few that we didn't test, so it would be nice to add tests for them (to check that we've added the checks for both CPU and CUDA:

For: TensorIterator::unary_op
- elu: test in-place variant: https://github.com/pytorch/pytorch/blob/2b7108a96f3de4f20a18b44f2f64ab0e7ded3ca5/aten/src/ATen/native/Activation.cpp#L111-L117
- hardswish: test in-place variant: https://github.com/pytorch/pytorch/blob/2b7108a96f3de4f20a18b44f2f64ab0e7ded3ca5/aten/src/ATen/native/Activation.cpp#L150-L154 
- leaky_relu: test in-place variant https://github.com/pytorch/pytorch/blob/2b7108a96f3de4f20a18b44f2f64ab0e7ded3ca5/aten/src/ATen/native/Activation.cpp#L713-L735

For TensorIterator::binary_op:
- index_add_: https://github.com/pytorch/pytorch/blob/2b7108a96f3de4f20a18b44f2f64ab0e7ded3ca5/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L375 has two paths.
- atan2: test out variant and in-place variant: https://github.com/pytorch/pytorch/blob/2b7108a96f3de4f20a18b44f2f64ab0e7ded3ca5/aten/src/ATen/native/BinaryOps.cpp#L337-L350 
- ilshift: https://github.com/pytorch/pytorch/blob/2b7108a96f3de4f20a18b44f2f64ab0e7ded3ca5/aten/src/ATen/native/BinaryOps.cpp#L596-L607 
- irshift: https://github.com/pytorch/pytorch/blob/2b7108a96f3de4f20a18b44f2f64ab0e7ded3ca5/aten/src/ATen/native/BinaryOps.cpp#L624-L635
- hypot: test out variant and in-place variant: https://github.com/pytorch/pytorch/blob/2b7108a96f3de4f20a18b44f2f64ab0e7ded3ca5/aten/src/ATen/native/BinaryOps.cpp#L888-L903
- nextafter: test out variant and in-place variant: https://github.com/pytorch/pytorch/blob/2b7108a96f3de4f20a18b44f2f64ab0e7ded3ca5/aten/src/ATen/native/BinaryOps.cpp#L905-L920
- threshold: test in-place variant: https://github.com/pytorch/pytorch/blob/2b7108a96f3de4f20a18b44f2f64ab0e7ded3ca5/aten/src/ATen/native/Activation.cpp#L380-L392

zou3519 replied:
In retrospect, because there are so many ops, I think this PR would be easier to review if we split it up into, e.g.:
- a PR to remove TensorIterator::unary_op's check_mem_overlap (and instead default to True)
- a PR to remove TensorIterator::binary_op's check_mem_overlap (and instead default to True)
- ...
- a PR that hits the switch for the default of `check_mem_overlap` in TensorIterator to True.

However, I'm fine with reviewing all the changes in one single PR (this PR) because the ultimate line count will probably still be manageable.

zou3519 commented on aten/src/ATen/native/TensorIterator.h:
```diff
@@ -553,7 +550,7 @@ class CAFFE2_API TensorIteratorConfig final {
   c10::optional<DimVector> static_shape_ = c10::nullopt;
   c10::optional<std::pair<ScalarType, Device>> static_dtype_and_device_ = c10::nullopt;
-  bool check_mem_overlap_ = false;
+  bool check_mem_overlap_ = true;
```
We should add some prescriptive documentation to https://github.com/pytorch/pytorch/blob/2b7108a96f3de4f20a18b44f2f64ab0e7ded3ca5/aten/src/ATen/native/TensorIterator.h#L429-L432 describing when a user should `set_check_mem_overlap = False`, and that the default is True.

zou3519 reviewed (commented):
(not a full review, I'm still going through the ops)

peterbell10 commented:
Closing as this was superseded by gh-43423
